### PR TITLE
CVE-2013-6649 and CVE-2015-1217

### PR DESCRIPTION
**Pull request by AlexanderKidd**
Filled out details for the respective CVE yamls.

### CODE REVIEW

File: cves/CVE-2013-6649.yml [hunk 1]
```diff
@@ -15,7 +15,15 @@ description_instructions: |
   that outsiders to Chromium would not understand. Technology like "regular
   expressions" is fine, and security phrases like "invalid write" are fine to
   keep too.
-description: 
+description: |
+  A use-after-free vulnerability in the RenderSVGImage::paint function in the
```
Comment (andymeneely) [inline cves/CVE-2013-6649.yml]: Don't mention function names in this opening statement. Too much jargon. Work on telling this story to someone.

File: cves/CVE-2013-6649.yml [hunk 2]
```diff
@@ -27,22 +35,28 @@ reviews:
...
+    catch for edge cases like a zero-size image.
 upvotes_instructions: |
   For the first round, ignore this upvotes number.
 
   For the second round of reviewing, you will be giving a certain amount of
   upvotes to each vulnerability you see. Your peers will tell you how
   interesting they think this vulnerability is, and you'll add that to the
   upvotes score on your branch.
-upvotes: 
+upvotes:
```
Comment (andymeneely) [inline cves/CVE-2013-6649.yml]: Give this one 2 upvotes from me.

File: cves/CVE-2013-6649.yml [hunk 3]
```diff
@@ -173,4 +222,16 @@ mistakes:
...
+    if statements which make it hard sometimes to accurately place validation
+    conditions.  Checking for null/empty values is a must, especially in a
+    low-level language like C/C++ which is what Chromium is written in.  These
+    kinds of mistakes are commonly associated with use-after-free and buffer
+    overflows due to memory mismanagement.  It is interesting to note tests
+    were added, but none of them appear to directly check for zero-size images.
+
+    The code was also reviewed on that VCC commit, so reviewing procedures should
+    probably be revised, especially to review edge cases and validation in and
+    around the changed code.
```
Comment (andymeneely) [inline cves/CVE-2013-6649.yml]: Excellent!! I'm labeling this PR as a good example for others to follow. Please take my feedback in, but this is very much what I'm looking for. Thank you!

File: cves/CVE-2013-6649.yml (hunk 2 above)
Comment (andymeneely) [inline cves/CVE-2013-6649.yml]: Nah make it 5 upvotes from me.

File: cves/CVE-2015-1217.yml [hunk 4]
```diff
@@ -15,7 +15,22 @@ description_instructions: |
   that outsiders to Chromium would not understand. Technology like "regular
   expressions" is fine, and security phrases like "invalid write" are fine to
   keep too.
-description: 
+description: |
+  The V8LazyEventListener listens for javascript events for a given type of event.
```
Comment (andymeneely) [inline cves/CVE-2015-1217.yml]: Good - this is not so jargony because it explains it.

File: cves/CVE-2015-1217.yml [hunk 5]
```diff
@@ -15,7 +15,22 @@ description_instructions: |
...
+
+  The vulnerability category is "type confusion", where a variable or object is
+  previously declared as one type but is accessed as another type.  This causes
+  problems in mainly C-like languages because they do not have memory safety that
+  will check if something like a pointer is trying to point to allocated memory
+  of another type.  In this case, JavaScript code is constructed, compiled, and
+  then ran, partially based on DOM element input (what the listener should listen for).
+  The result is then cast to a function, although it is not necessarily a function,
+  causing type confusion that could cause a crash in the program or another DoS of
+  sorts for a malicious actor.
```
Comment (andymeneely) [inline cves/CVE-2015-1217.yml]: Excellent!! Can you briefly describe how it was fixed? Third paragraph, like a sentence or two.

File: cves/CVE-2015-1217.yml [hunk 6]
```diff
@@ -32,22 +47,37 @@ reviews:
...
   CVE-2011-3092.yml). Fixes and VCCs follow the same format.
 fixes:
 - :commit: fc81fcf38edd250876cc384a6ed5567e1b2999e4
-  :note: ''
-vccs: []
+  :note: |
+    Asserts were replaced with conditional statements in the code for
+    checking whether a context object is a document, whether it allows
+    inline event handlers, and whether the internal script that is ran
+    has a result that is a function.
```
Comment (andymeneely) [inline cves/CVE-2015-1217.yml]: Yeah just a quick summary of this sentence in the description would be fine.

File: cves/CVE-2015-1217.yml [hunk 7]
```diff
@@ -32,22 +47,37 @@ reviews:
...
+    the fix.
 upvotes_instructions: |
   For the first round, ignore this upvotes number.
 
   For the second round of reviewing, you will be giving a certain amount of
   upvotes to each vulnerability you see. Your peers will tell you how
   interesting they think this vulnerability is, and you'll add that to the
   upvotes score on your branch.
-upvotes: 
+upvotes:
```
Comment (andymeneely) [inline cves/CVE-2015-1217.yml]: 10 upvotes from me.

File: cves/CVE-2015-1217.yml [hunk 8]
```diff
@@ -78,20 +111,26 @@ discovered:
...
-  google: 
-  contest: 
+  answer: |
+    The renderer crashed when using a certain test page, which had a certain mouse
+    down event on the document.  Asserts were not enabled on the Windows Vista snapshot,
+    but the reporter, pim...@live.nl was able to debug that the preparedListenerObject
+    function had a type confusion vulnerability that caused the crash.
+  date: 2015-02-06
+  automated: true
+  google: false
```
Comment (andymeneely) [inline cves/CVE-2015-1217.yml]: Can you look into the person who discovered this? Maybe they have a LinkedIn profile or other information about what they do.

File: cves/CVE-2015-1217.yml [hunk 9]
```diff
@@ -100,10 +139,11 @@ interesting_commits:
...
   commits:
-  - commit: 
-    note: 
-  - commit: 
-    note: 
+  - commit: 1e2aefdb3d864e1bfd2e6e1c94c52aebe321c2d1
+    note: |
+      This is actually another bug involving a null state with the AbstractEventListener
+      in v8.  Could the developers had included looking at similar components such
+      as the LazyEventListener to find similar problems during the code review?
```
Comment (andymeneely) [inline cves/CVE-2015-1217.yml]: A good example of what I'm looking for. How much earlier was this from the fix?

File: cves/CVE-2015-1217.yml [hunk 10]
```diff
@@ -112,12 +152,18 @@ major_events:
 
     The event doesn't need to be directly related to this vulnerability, rather,
     we want to capture what the development team was dealing with at the time.
-  answer: 
+  answer: |
+    Not many major commits/events were created near this time, but over the past year
+    before this, the following was an interesting focus the team had:
```
Comment (andymeneely) [inline cves/CVE-2015-1217.yml]: End with a period.

File: cves/CVE-2015-1217.yml [hunk 11]
```diff
@@ -134,38 +180,50 @@ lessons:
...
-    note: 
+    applies: true
+    note: |
+      Judging from this string of commits for this file and others,
+      a strong sense of "you ain't gonna need it" seems to be present
+      with incremental but sometimes verbose additions of code that
+      are either not accompanied with new test files or are not reviewed
+      carefully for common edge cases that could lead to vulnerabilities
+      (e.g., null checks).  When in doubt, it may be a good idea to
+      implement these simple checks.
```
Comment (andymeneely) [inline cves/CVE-2015-1217.yml]: Good!

File: cves/CVE-2015-1217.yml [hunk 12]
```diff
@@ -178,4 +236,17 @@ mistakes:
...
+    fixing this vulnerability.  More rigorous code reviews for common security standards
+    (especially pertaining to C and how these subsystems are typically implemented in
+    Chromium) could add up in the long run to have a more fortified application build
+    with fewer common, easy to check for errors.  Make sure everyone reviewing understands
+    what the code is trying to accomplish and if it does that (Validate and Verify).
+
+    The fix of checking for certain conditions before preceding to create the listener
+    object seems appropriate for this kind of vulnerability.  In the future, do not
+    rely too heavily on asserts for runtime in the production environment.  These should
+    be converted to actual checks and handling of errors and exceptions.
```
Comment (andymeneely) [inline cves/CVE-2015-1217.yml]: Excellent!!

File: cves/CVE-2013-6649.yml (hunk 2 above)
Comment (tlr3552) [inline cves/CVE-2013-6649.yml]: I'll give my remaining 6 upvotes to this CVE. Edge cases (if not done in unit tests) needs more recognition IMO, especially after seeing some dev teams' practices in my own reviews...

File: cves/CVE-2013-6649.yml [hunk 13]
```diff
@@ -55,9 +69,16 @@ unit_tested:
 
     For the "fix" answer below, check if the fix for the vulnerability involves
     adding or improving an automated test to ensure this doesn't happen again.
-  answer: 
-  code: 
-  fix: 
+  answer: |
```
Comment (tlr3552) [inline cves/CVE-2013-6649.yml]: Your description is good, this bug must have been well documented with detailed origin information.

File: cves/CVE-2013-6649.yml [hunk 14]
```diff
@@ -55,9 +69,16 @@ unit_tested:
...
+    Automated unit tests were involved.  Specifically, an assertion failure was
+    thrown for the stack size equalling one and the heap having a use-after-free
+    error.  The VCC even had tests added to test the layout of the bufferd SVG
+    that was rendered where the vulnerability occurred.
+
+    The ASAN (Address Sanitizer) reported this bug, caught in the
+    WebCore::ScrollView::paint function.
```
Comment (tlr3552) [inline cves/CVE-2013-6649.yml]: This comment including system specific language will like be more confusing than helpful for the examiners of our research.

File: cves/CVE-2015-1217.yml (hunk 6 above)
Comment (AlexanderKidd) [inline cves/CVE-2015-1217.yml]: Good thing I nearly copy-pasta'd it.

File: cves/CVE-2013-6649.yml [hunk 15]
```diff
@@ -27,22 +35,28 @@ reviews:
...
 fixes:
 - :commit: 70bcb6b3396a395e871e10b2ff883d92b8218e9f
-  :note: ''
-vccs: []
+  :note: |
+    If the SVG image is of zero size, if condition does not execute rendering the image.
+vccs:
+- :commit: c4df847fabcf7e172dd0395708a9d93385050c8f
+  :note: |
+    Arguably the first and only VCC.  A substantial chunk of code was added to bufferd
```
Comment (rjm6823) [inline cves/CVE-2013-6649.yml]: Just a wording comment, I feel like you should commit to either this being the only VCC or also include another commit that could be a VCC.

File: cves/CVE-2013-6649.yml [hunk 16]
```diff
@@ -27,22 +35,28 @@ reviews:
...
+    catch for edge cases like a zero-size image.
 upvotes_instructions: |
   For the first round, ignore this upvotes number.
 
   For the second round of reviewing, you will be giving a certain amount of
   upvotes to each vulnerability you see. Your peers will tell you how
   interesting they think this vulnerability is, and you'll add that to the
   upvotes score on your branch.
-upvotes: 
+upvotes: 11
```
Comment (rjm6823) [inline cves/CVE-2013-6649.yml]: +2 Upvotes for me, rendering nothing is always a tricky task.
